File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseContainsKey.vb
Imports System.Collections.Generic
Imports System.Linq

Class Compliant
    Function UsesContainsKey() As Boolean
        Dim dict = New Dictionary(Of Integer, Integer) ' Compliant
        Return dict.ContainsKey(42)
    End Function

    Function KeysNotFollowedByContains() As Boolean
        Dim dict = New Dictionary(Of Integer, Integer) ' Compliant
        Return dict.Keys.Any()
    End Function

    Function KeysNotFollowedByAnything() As IEnumerable(Of Integer)
        Dim dict = New Dictionary(Of Integer, Integer) ' Compliant
        Return dict.Keys
    End Function

    Function NoDictionary() As Boolean
        Dim dict = New NoDictionary()
        Return dict.Keys.Contains(42) ' Compliant
    End Function
End Class

Class NonCompliant
    Function KeysContains() As Boolean
        Dim dict = New Dictionary(Of Integer, Integer) ' NonCompliant {{Use ContainsKey() instead.}}
        Return dict.Keys.Contains(42)
        '                ^^^^^^^^
    End Function
End Class

Class NoDictionary
    Public ReadOnly Property Keys As List(Of Integer)
End Class

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseContainsKey.cs
using System.Collections.Generic;
using System.Linq;

class Compliant
{
    bool UsesContainsKey()
    {
        var dict = new Dictionary<int, int>();
        return dict.ContainsKey(42); // Compliant
    }

    bool KeysNotFollowedByContains()
    {
        var dict = new Dictionary<int, int>();
        return dict.Keys.Any(); // Compliant
    }

    IEnumerable<int> KeysNotFollowedByAnything()
    {
        var dict = new Dictionary<int, int>();
        return dict.Keys; // Compliant
    }

    bool NoDictionary()
    {
        var dict = new NoDictionary();
        return dict.Keys.Contains(42); // Compliant
    }
}

class NonCompliant
{
    bool KeysContains()
    { 
        var dict = new Dictionary<int, int>();
        return dict.Keys.Contains(42); // NonCompliant {{Use ContainsKey() instead.}}
        //               ^^^^^^^^
    }
}

class NoDictionary
{
    public List<int> Keys { get; }
}
